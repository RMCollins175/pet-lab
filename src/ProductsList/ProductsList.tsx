interface ProductsType {
  products: ProductType[];
}

export interface ProductType {
  id: number;
  slug: string;
  title: string;
  vendor: string;
  tags: Array<string>;
  published: true;
  url: string;
  image_src: string;
  option_value: string;
  sku: string;
  price: number;
  subscription_discount: number;
  subscription: true;
}

export const ProductsList = ({ products }: ProductsType) => {
  return (
    <ul className="users-list">
      {products.map((product: ProductType) => (
        <li
          key={product.id}
          className="product-item"
          data-testID="product-list"
        >
          <span data-testID="product-name">{product.title}</span>
        </li>
      ))}
    </ul>
  );
};
