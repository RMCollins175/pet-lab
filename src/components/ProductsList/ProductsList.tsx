interface ProductsListProps {
  products: ProductType[];
}

export interface ProductType {
  id: number;
  slug: string;
  title: string;
  vendor: string;
  tags: Array<string>;
  published: true;
  url: string;
  image_src: string;
  option_value: string;
  sku: string;
  price: number;
  subscription_discount: number;
  subscription: true;
}

export const ProductsList = ({ products }: ProductsListProps) => {
  const getDiscount = (discount: any) => {
    return typeof discount === "number" ? `${discount}%` : "";
  };

  return (
    <ul className="products-list">
      <li className="product-item" data-testID="product-list">
        <div style={{ paddingLeft: "50px" }} />
        <span data-testID="product-name">
          <strong>Title</strong>
        </span>
        <span data-testID="product-price">
          <strong>Price</strong>
        </span>
        <span data-testID="subscription-discount">
          <strong>Discount</strong>
        </span>
      </li>
      {/* TODO make into Header component */}
      {products.map((product: ProductType) => (
        <li
          key={product.id}
          className="product-item"
          data-testID="product-list"
        >
          <img src={product.image_src} alt={product.title} height={50} />
          <span data-testID="product-name">{product.title}</span>
          <span data-testID="product-price">Â£{product.price}</span>
          <span data-testID="subscription-discount">
            {getDiscount(product.subscription_discount)}
          </span>
        </li>
      ))}
    </ul>
  );
};
