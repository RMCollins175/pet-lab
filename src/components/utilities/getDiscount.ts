export const getDiscount = (discount: number) => {
  return typeof discount === "number" ? `${discount}%` : "";
};
