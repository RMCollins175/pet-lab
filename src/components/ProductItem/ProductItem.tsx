import { ProductType } from "../ProductsList/ProductsList";
import { getDiscount } from "../utilities/getDiscount";

interface ProductProps {
  product: ProductType;
}

export const ProductItem = ({ product }: ProductProps) => {
  return (
    <li
      className="grid justify-center grid-cols-[0.2fr_1.5fr_0.5fr_0.5fr] p-4 border-b border-gray-300 cursor-pointer "
      data-testid="product-list">
      <img src={product.image_src} alt={product.title} height={50} />
      <span
        className="flex items-center justify-center"
        data-testid="product-name">
        {product.title}
      </span>
      <span
        className="flex items-center justify-center"
        data-testid="product-price">
        Â£{product.price}
      </span>
      <span
        className="flex items-center justify-center"
        data-testid="subscription-discount">
        {getDiscount(product.subscription_discount)}
      </span>
    </li>
  );
};
