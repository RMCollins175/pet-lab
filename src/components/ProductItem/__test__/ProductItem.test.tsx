import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ProductItem } from "../ProductItem";
import { getDiscount } from "../../utilities/getDiscount";
import { ProductType } from "../../ProductsList/ProductsList";

describe("ProductItem Component", () => {
  const mockProduct: ProductType = {
    id: 1,
    slug: "test-product",
    title: "Test Product",
    vendor: "Test Vendor",
    tags: ["tag1", "tag2"],
    published: true,
    url: "http://example.com",
    image_src: "test-image.jpg",
    option_value: "option1",
    sku: "SKU123",
    price: 20,
    subscription_discount: 10,
    subscription: true,
  };

  test("renders product information correctly", () => {
    render(<ProductItem product={mockProduct} />);
    expect(screen.getByText(mockProduct.title)).toBeInTheDocument();
    expect(screen.getByText(`Â£${mockProduct.price}`)).toBeInTheDocument();
    expect(
      screen.getByText(getDiscount(mockProduct.subscription_discount))
    ).toBeInTheDocument();
    expect(screen.getByAltText(mockProduct.title)).toHaveAttribute(
      "src",
      mockProduct.image_src
    );
  });

  test("renders an image with correct attributes", () => {
    render(<ProductItem product={mockProduct} key={mockProduct.id} />);

    const image = screen.getByAltText(mockProduct.title);
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", mockProduct.image_src);
    expect(image).toHaveAttribute("height", "50");
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ProductItem product={mockProduct} key={mockProduct.id} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
