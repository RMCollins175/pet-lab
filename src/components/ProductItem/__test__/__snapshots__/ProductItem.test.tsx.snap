// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Component matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="grid justify-center grid-cols-[0.2fr_1.5fr_0.5fr_0.5fr] p-4 border-b border-gray-300 cursor-pointer "
    data-testid="product-list"
  >
    <img
      alt="Test Product"
      height="50"
      src="test-image.jpg"
    />
    <span
      class="flex items-center justify-center"
      data-testid="product-name"
    >
      Test Product
    </span>
    <span
      class="flex items-center justify-center"
      data-testid="product-price"
    >
      Â£20
    </span>
    <span
      class="flex items-center justify-center"
      data-testid="subscription-discount"
    >
      10%
    </span>
  </li>
</DocumentFragment>
`;
